### PR TITLE
[GSheets] Add new app for OAuth Integration

### PR DESCRIPTION
**Pull request by RynhardPietersen**
## Product Description

This is a new branch because the old one was getting cluttered with unnecessary commits. [link](https://github.com/dimagi/commcare-hq/pull/30982)

We want to include a new option for users to export reports to google sheets. It would not make sense for most of the code that interfaces with google to live in exports as this code is mostly for redirecting to google for authorization and handling the credentials that we receive from google. As such we have decided to create a new oauth app for this code to live in and also for future oauth endeavours.

## Technical Summary

link: https://dimagi-dev.atlassian.net/browse/SS-272

I've decided to create view functions instead of view classes because these functions will not be handling any user facing code. They are simply there to handle any code that needs to run before redirecting to Google for login (such as checking whether a user has any stored credentials already) and for handling data that we receive from Google after a user has given us access to their Google account (such as storing or updating credentials).

[Here](https://github.com/dimagi/commcare-cloud/pull/5145/files) is a pr to some changes that needed to be made in commcare-cloud

## Feature Flag

This should probably be hidden behind the Google Sheets feature flag, but seeing as none of this code will be user facing, I feel that it is unnecessary.

## Safety Assurance

None of the code is user facing. I have written and ran tests for the code. Will upload to staging and see if it has any effect on staging. 

### Safety story

There is currently no way to access any of these changes through cchq as view that accesses these views is on a different branch. I have also done a lot of local testing and will upload this to staging.

### QA Plan

QA will be needed once the export view has been implemented

### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/oauth_integrations/utils.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+        if(row_number == 0):
+            headers = list(row_values.keys())
+            data.append(headers)
+        data.append(list(row_values.values()))
+
+    return data
+
+
+def chunkify_data(data, chunk_size):
```
Comment (biyeun) [inline corehq/apps/oauth_integrations/utils.py:95]: where is this actually being used? I don't see a reference to it (outside of tests) anywhere else in this PR

File: migrations.lock [hunk 2]
```diff
@@ -921,7 +923,6 @@ two_factor
  0004_auto_20160205_1827
  0005_auto_20160224_0450
  0006_phonedevice_key_default
- 0007_auto_20201201_1019
```
Comment (biyeun) [inline migrations.lock:924]: is the deletion of this migration intentional? did you `pip install` or `pip sync` requirements lately? this doesn't look related to the oauth work

**Review by biyeun — CHANGES_REQUESTED** [review]
left a few comments. some cleanup suggestions and then one about `migrations.lock` looking a little 🐟

File: migrations.lock (hunk 2 above)
Comment (RynhardPietersen) [reply migrations.lock:924]: I didn't delete this intentionally. I have done `pip install requirements` today

**Review by snopoke — COMMENTED** [review]
This seems very 'google' specific. Is there a reason it can't be more generic? Did we consider using [django-allauth](https://django-allauth.readthedocs.io/en/latest/overview.html) which already supports Google?

Also, how does this overlap with the oauth support we have for data forwarding: https://github.com/dimagi/commcare-hq/pull/27531

File: corehq/apps/oauth_integrations/tests/test_util.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+        stringified_credentials = stringify_credentials(self.credentials)
+        loaded_credentials = load_credentials(stringified_credentials)
+
+        self.assertEqual(loaded_credentials.token, desired_credentials.token)
+
+    def test_chunkify_data(self):
+        desired_list = ['This is a string of ', 'data that I want to ', 'break up in four chu', 'nks']
+        chunk_length = 20
+        data = "This is a string of data that I want to break up in four chunks"
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/oauth_integrations/tests/test_util.py:94]: yeah, that's what i was thinking. maybe better to remove this method for right now and wait until it's actually being used with real data. I think you may have some issues chunking the data this way as it will break apart the stringified JSON in ways that are not desirable

**Comment by RynhardPietersen** [comment]
@snopoke I'm sorry for the delayed response.

Firstly, I would like to say that my current work with oauth2 as seen here is very google-centric as it pertains to the Google Sheet integration currently under development.

Unfortunately we did not consider django-allauth, mainly because it wasn't recommended by anyone that reviewed the spec document. I have taken a look at django-allauth and I see why we would want to use it, but because of how google-centric the oauth is right now, and because of how far in development I am, I feel that it would be better to stay the course.

I closely followed Google Documentation to interface with their api.

Hopefully this answers all of your questions :)

**Comment by snopoke** [comment]
> Unfortunately we did not consider django-allauth, mainly because it wasn't recommended by anyone that reviewed the spec document. I have taken a look at django-allauth and I see why we would want to use it, but because of how google-centric the oauth is right now, and because of how far in development I am, I feel that it would be better to stay the course.

very unfortunate :(

**Comment by snopoke** [comment]
> it wasn't recommended by anyone that reviewed the spec

Can you link to the spec please. I don't see it in the Jira ticket or the PR description.

**Comment by RynhardPietersen** [comment]
[Here](https://docs.google.com/document/d/1_87adUGD__uPz8LnUsI0jOOuq80o2k0rB0tIjU1Utmg/edit#heading=h.qcr9htyocy8i) you go

**Comment by biyeun** [comment]
hey @snopoke I spoke with Kirti today, and we are going to incorporate cross-division tech leads as part of spec reviews moving forward. I'm sorry that this slipped by without you getting an opportunity to review and comment at the speccing stage. We want to ensure that going forward we can gather all the appropriate inputs at the right stages.

I think due to time constraints it would add some time for Rynhard to change course right now, and it doesn't seem super practical to do so as there isn't a future plan to integrate with other social platforms. I think if that requirement does change, this work can be easily migrated to use `django-allauth`.

**Review by biyeun — APPROVED** [review]
🎉
